compiler/src/dotty/tools/dotc/transform/PatternMatcher.scala

```diff
@@ -1,1871 +1,931 @@
-import dotty.tools.dotc.core.{TypeApplications, Flags}
-import dotty.tools.dotc.typer.Applications
-import dotty.tools.dotc.util.Positions
-import typer.ErrorReporting._
+import collection.mutable
+import SymDenotations._, Symbols._, Contexts._, Types._, Names._, StdNames._, NameOps._
 import ast.Trees._
-import Applications._
-import TypeApplications._
-import SymUtils._, core.NameOps._
-import core.Mode
-import patmat._
-
-import dotty.tools.dotc.util.Positions.Position
-import dotty.tools.dotc.core.Decorators._
-import dotty.tools.dotc.core.Flags
-
-/** This phase rewrites pattern matches.
- *  FIXME: A more detailed explanation would be good.
+import util.Positions._
+import typer.Applications.{isProductMatch, isGetMatch, productSelectors}
+import SymUtils._
+import Flags._, Constants._
+import Decorators._
+import patmat.Space
+import NameKinds.{UniqueNameKind, PatMatStdBinderName, PatMatCaseName}
+import config.Printers.patmatch
+
+/** The pattern matching transform.
+ *  After this phase, the only Match nodes remaining in the code are simple switches
+ *  where every pattern is an integer constant
  */
-class PatternMatcher extends MiniPhaseTransform with DenotTransformer {
-  import dotty.tools.dotc.ast.tpd._
-
-  override def transform(ref: SingleDenotation)(implicit ctx: Context): SingleDenotation = ref
+class PatternMatcher extends MiniPhaseTransform {
+  import ast.tpd._
+  import PatternMatcher._
 
+  override def phaseName = "patternMatcher"
   override def runsAfter = Set(classOf[ElimRepeated])
-
   override def runsAfterGroupsOf = Set(classOf[TailRec]) // tailrec is not capable of reversing the patmat tranformation made for tree
 
-  override def phaseName = "patternMatcher"
-
-  private var _id = 0 // left for debuging
-
   override def transformMatch(tree: Match)(implicit ctx: Context, info: TransformerInfo): Tree = {
-    val translated = new Translator()(ctx).translator.translateMatch(tree)
+    val translated = new Translator(tree.tpe, this).translateMatch(tree)
 
     // check exhaustivity and unreachability
-    val engine = new SpaceEngine
+    val engine = new patmat.SpaceEngine
+
     if (engine.checkable(tree)) {
       engine.checkExhaustivity(tree)
       engine.checkRedundancy(tree)
     }
 
     translated.ensureConforms(tree.tpe)
   }
+}
 
-  class Translator(implicit ctx: Context) {
-
-  def translator = {
-    new OptimizingMatchTranslator/*(localTyper)*/
-  }
-
-  class OptimizingMatchTranslator extends MatchOptimizer/*(val typer: analyzer.Typer)*/ with MatchTranslator
-
-  trait CodegenCore {
-
-    // assert(owner ne null); assert(owner ne NoSymbol)
-    def freshSym(pos: Position, tp: Type = NoType, unique: UniqueNameKind = PatMatStdBinderName, owner: Symbol = ctx.owner) = {
-      ctx.newSymbol(owner, unique.fresh(), Flags.Synthetic | Flags.Case, tp, coord = pos)
+object PatternMatcher {
+  import ast.tpd._
+
+  final val selfCheck = true // debug option, if on we check that no case gets generated twice
```

Comment:
No, I think we can drop it. On would typically enable it if something funky happens in the backend with double definitions.